src/igraph/constructors/from_object.py

```diff
@@ -0,0 +1,462 @@
+            try:
+                edge_attributes[name].append(item[index])
+            except IndexError:
+                edge_attributes[name].append(None)
+
+    # Set up the "name" vertex attribute
+    vertex_attributes = {}
+    vertex_attributes[vertex_name_attr] = list(idgen.values())
+    n = len(idgen)
+
+    # Construct the graph
+    return cls(n, edge_list, directed, {}, vertex_attributes, edge_attributes)
+
+
+def construct_graph_from_sequence_dict(
+    cls,
+    edges,
+    directed=False,
+    vertex_name_attr="name",
+):
+    """Constructs a graph from a dict-of-sequences representation.
+
+    This function is used to construct a graph from a dictionary of
+    sequences (e.g. of lists). For each key x, its corresponding value is
+    a sequence of multiple objects: for each y, the edge (x,y) will be
+    created in the graph. x and y must be either one of:
+
+    - two integers: the vertices with those ids will be connected
+    - two strings: the vertices with those names will be connected
+
+    If names are used, the order of vertices is not guaranteed, and each
+    vertex will be given the vertex_name_attr attribute.
+    """
+    # Check if we are using names or integers
+    for item in edges:
+        break
+    if not isinstance(item, (int, str)):
+        raise ValueError("Keys must be integers or strings")
+
+    vertex_attributes = {}
+    if isinstance(item, str):
+        name_map = UniqueIdGenerator()
+        edge_list = []
+        for source, sequence in edges.items():
+            source_id = name_map[source]
+            for target in sequence:
+                edge_list.append((source_id, name_map[target]))
+        vertex_attributes[vertex_name_attr] = name_map.values()
+        n = len(name_map)
+
+    else:
+        edge_list = []
+        n = 0
+        for source, sequence in edges.items():
+            n = max(n, source, *sequence)
+            for target in sequence:
+                edge_list.append((source, target))
+        n += 1
+
+    # Construct the graph
+    return cls(n, edge_list, directed, {}, vertex_attributes, {})
+
+
+def construct_graph_from_dict_dict(
+    cls,
+    edges,
+    directed=False,
+):
+    """Constructs a graph from a dict-of-dicts representation.
+
+    Each key can be an integer or a string and represent a vertex. Each value
+    is a dict representing edges (outgoing if the graph is directed) from that
+    vertex. Each dict key is an integer/string for a target vertex, such that
+    an edge will be created between those two vertices, while each value is
+    a dictionary of edge attributes for that edge.
+
+    Example:
+
+    {'Alice': {'Bob': {'weight': 1.5}, 'David': {'weight': 2}}}
+
+    creates a graph with three vertices (Alice, Bob, and David) and two edges:
+    
+    - Alice - Bob (with weight 1.5)
+    - Alice - David (with weight 2)
+    """
+    # Check if we are using names or integers
+    for item in edges:
```

Comment:
Same here; this won't work if `edges` is empty because `item` will be unassigned.

Reply:
yep, fixed now